### ISSUE TITLE
Readonly mode for database

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-framework#5693]
**As a sysadmin, I want to have a way of setting SilverStripe to readonly mode, so that I can safely reduce downtime during deployments**
# Acceptance criteria
- Readonly mode can be activated by a config setting or an `_ss_environment.php` define.
- When in readonly mode, the ORM returns sensible defaults about locking off edit rights to various features.
- Where non-critical database edits occur (e.g. logging of log-ins), they can be silently skipped. Developers have a way of defining such cases.
- Ensure that at the low level of the DB, writes and schema changes are blocked. `DB::query("INSERT ...")` shouldn't bypass this readonly-mode.
- Where code hasn't been written to handle readonly mode, exceptions should be thrown. A config option exists to ignore all errors instead.
- Dev docs are written

**Comment by fspringveldt** [comment]
I've created a module which checks 5/6 of the boxes and will continue working on it until 6/6 is achieved at https://github.com/fspringveldt/db-read-only-mode for review.

**Comment by sminnee** [comment]
@tractorcow Franco's module looks good—you should check it out. :-)

**Comment by tractorcow** [comment]
I've done a code review and it's excellent. :)

Take care: DBSchemaManager::query() will still invoke query on the UN-mocked database service, so make sure that your proxy injects itself into the schema instance as well.

**Comment by fspringveldt** [comment]
Thanks for the review and suggestion Damian.

I'll inject DBSchemaManager as well and write a test for it too.

Regards,
Franco

**Comment by dhensby** [comment]
This has missed 3.5 and can't go into a new patch release. I highly doubt a 3.6 release will happen so this needs to target master. sorry.

**Comment by tractorcow** [comment]
We might but probably won't have 3.6.

**Comment by sminnee** [comment]
> highly doubt a 3.6 release will happen so this needs to target master. sorry.

I disagree with this. SilverStripe 3.x is going to be supported until Dec 2018.

This is good for 3 and was posted here at my request.

**Comment by GuySartorelli** [comment]
Closing - we have our hands full and nobody has been asking for this that I've heard of.
If there's a lot of interest we can reopen
